### PR TITLE
8334870: javac does not accept classfiles with certain permitted RuntimeVisibleParameterAnnotations and RuntimeInvisibleParameterAnnotations attributes

### PR DESCRIPTION
**Pull request by lahodaj**
JVMS 4.7.18 and 4.7.19 say this about the entries in the RuntimeVisibleParameterAnnotations and RuntimeInvisibleParameterAnnotations attributes:

> The i'th entry in the parameter_annotations table may, but is not required to, correspond to the i'th parameter descriptor in the method descriptor (§4.3.3).

When reading classfiles, javac does not follow this specification (which puts no requirements on the number of entries in the attributes), but rather expects as many entries as the `MethodSymbol.type` has (which is either based in the `Signature` attribute, if present, or the method descritor, if `Signature` is not present).

This patch proposes to adjust javac to accept the attributes with any number of entries, implementing a number of heuristics to map the content of the entries in the attributes to the parameters. There are comments in the code explaining the heuristics.

javac will ignore the RuntimeVisibleParameterAnnotations and RuntimeInvisibleParameterAnnotations attributes it cannot handle with a warning.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8334871](https://bugs.openjdk.org/browse/JDK-8334871) to be approved

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8334870: javac does not accept classfiles with certain permitted RuntimeVisibleParameterAnnotations and RuntimeInvisibleParameterAnnotations attributes`

### Issues
 * [JDK-8334870](https://bugs.openjdk.org/browse/JDK-8334870): javac does not accept classfiles with certain permitted RuntimeVisibleParameterAnnotations and RuntimeInvisibleParameterAnnotations attributes (**Bug** - P4)
 * [JDK-8334871](https://bugs.openjdk.org/browse/JDK-8334871): javac does not accept classfiles with certain permitted RuntimeVisibleParameterAnnotations and RuntimeInvisibleParameterAnnotations attributes (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19860/head:pull/19860` \
`$ git checkout pull/19860`

Update a local copy of the PR: \
`$ git checkout pull/19860` \
`$ git pull https://git.openjdk.org/jdk.git pull/19860/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19860`

View PR using the GUI difftool: \
`$ git pr show -t 19860`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19860.diff">https://git.openjdk.org/jdk/pull/19860.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19860#issuecomment-2186813523)

### CODE REVIEW

File: test/langtools/tools/javac/annotations/parameter/ParameterAnnotations.java [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+        super(System.err);
+        tb = new ToolBox();
+    }
+
+    public void runTests() throws Exception {
+        runTests(m -> new Object[] { Paths.get(m.getName()) });
+    }
+
+    @Test
+    public void testEnum(Path base) throws Exception {
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/annotations/parameter/ParameterAnnotations.java:87]: I think that for completeness, we should add a test for a record with a compact constructor as all the parameters of that constructor will be mandated

File: test/langtools/tools/javac/annotations/parameter/ParameterAnnotations.java [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.util.ElementFilter;
+
+import toolbox.TestRunner;
+import toolbox.JavacTask;
+import toolbox.Task;
+import toolbox.ToolBox;
+
+public class ParameterAnnotations extends TestRunner {
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/annotations/parameter/ParameterAnnotations.java:69]: very nice test

File: test/langtools/tools/javac/annotations/parameter/ParameterAnnotations.java [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+
+    private MethodTransform NO_SIGNATURE = (builder, element) -> {
+        if (element instanceof SignatureAttribute) {
+            //ignore
+        } else {
+            builder.with(element);
+        }
+    };
+
+    private MethodTransform NO_SIGNATURE_NO_METHOD_PARAMETERS = (builder, element) -> {
```
Comment (liach) [inline test/langtools/tools/javac/annotations/parameter/ParameterAnnotations.java:449]:
```suggestion
    private MethodTransform NO_SIGNATURE_NO_METHOD_PARAMETERS = MethodTransform.dropping(element -> element instanceof SignatureAttribute ||
            element instanceof MethodParametersAttribute);
```
Shortcuts like this.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm, thanks for the updates

**Review by vicente-romero-oracle — APPROVED** [review]
last changes looks good

**Pull request closed without merging**